onnxruntime/core/optimizer/transpose_optimization/onnx_transpose_optimization.cc

```diff
@@ -245,32 +245,232 @@ static std::unique_ptr<api::NodeRef> GetDQIfProducingValue(const api::GraphRef&
+
+// Quantization modes for QuantizeLinear/DequantizeLinear.
+enum class QuantizationMode : uint8_t {
+  kUnknown,
+  kPerTensor,
+  kPerAxis,
+  kBlocked,
+};
+
+// Returns true if this optimizer supports the given quantization mode.
+constexpr bool IsSupportedQuantizationMode(QuantizationMode mode) {
+  return (mode == QuantizationMode::kPerTensor) || (mode == QuantizationMode::kPerAxis);
+}
+
+// Stores quantization info for a validated Q or DQ node.
+struct QuantizationInfo {
+  QuantizationMode mode;
+  int64_t norm_axis;  // 'normalized' axis is in the range [0, Input0Rank - 1]
+};
+
+// Returns the shape of a Q or DQ node's scale input, or std::nullopt if unable to get the shape.
+static std::optional<std::vector<int64_t>> GetQOrDQScaleShape(const api::GraphRef& graph,
+                                                              const api::NodeRef& q_or_dq_node) {
+  // Try to get the shape from ValueInfo.
+  std::string_view scale_input = q_or_dq_node.Inputs()[1];
+  const std::unique_ptr<api::ValueInfoRef> scale_value_info = graph.GetValueInfo(scale_input);
+  std::optional<std::vector<int64_t>> scale_shape = scale_value_info->Shape();
+  if (scale_shape) {
+    return scale_shape;
+  }
+
+  // ValueInfo is not available. Try to get shape from Tensor if scale is a constant.
+  const std::unique_ptr<api::TensorRef> dq_scale_constant = graph.GetConstant(scale_input);
+  if (dq_scale_constant) {
+    return dq_scale_constant->Shape();
+  }
+
+  return std::nullopt;
+}
+
+// Returns quantization information (quantization mode, normalized axis).
+// Returns std::nullopt if unable to get the quantization info or if the axis attribute is invalid.
+static std::optional<QuantizationInfo> GetQuantizationInfo(const api::GraphRef& graph,
+                                                           const api::NodeRef& q_or_dq_node) {
+  const std::vector<std::string_view> inputs = q_or_dq_node.Inputs();
+
+  // Need to use the scale input's shape to determine the quantization mode. Can't just use the presence of the axis
+  // attribute because even per-tensor Q/DQ have a default axis of 1.
+  std::optional<std::vector<int64_t>> scale_shape = GetQOrDQScaleShape(graph, q_or_dq_node);
+  if (!scale_shape) {
+    return std::nullopt;
+  }
+
+  QuantizationInfo quant_info = {};
+
+  if (IsScalarOr1Element1DTensor(*scale_shape)) {
+    // A scalar or tensor scale with shape (1,) indicates per-tensor quantization.
+    quant_info.mode = QuantizationMode::kPerTensor;
+    quant_info.norm_axis = 1;  // 1 is the default 'axis' even for per-tensor quantization (axis is ignored).
+  } else {
+    int64_t axis = q_or_dq_node.GetAttributeIntDefault("axis", 1);
+    const auto input0_info = graph.GetValueInfo(inputs[0]);
+    auto input0_rank = input0_info->ShapeRank();
+    if (!input0_rank.has_value() || !NormalizeAndValidateAxis(axis, *input0_rank)) {
+      return std::nullopt;  // Unable to normalize the DQ's axis.
```

Comment:
Side note: We really need to add some sort of log facility now that these subtle things are influencing the behavior. It was simpler in the beginning as there was no QDQ handling so it was trivial to see why a Transpose got added/blocked.

Reply:
Yeah, logging would probably help debugging/observability. I left a todo comment.